ls コマンドを作るオブジェクト指向版

heavy-stone opened the pull request:
課題提出用のコマンド
1
```
./ls.rb
```
```
\ls
```
2
```
./ls.rb -arl
```
```
\ls -arl
```
3
```
./ls.rb /no_exist1 /no_exist2 /tmp /usr 
```
```
\ls /no_exist1 /no_exist2 /tmp /usr 
```
4
```
./ls.rb -arl /no_exist1 /no_exist2 /tmp /usr
```
```
\ls -arl /no_exist1 /no_exist2 /tmp /usr 
```
5
```
rubocop
```
```
rake
```

heavy-stone committed:
add object oriented ls codes

heavy-stone committed:
add object oriented ls tests

betachelsea commented on 07.ls_object/lib/entry.rb:
```diff
@@ -0,0 +1,107 @@
+  end
+
+  def format_status_with_l_option(entry = self)
+    readlink_path = absolute_path(entry.path, entry.parent_path, entry.is_parent)
```
[nits] こちら、引数でもらってきたentryは自分自身ということで、もしかして `entry.absolute_path` のように使うことは可能だったりしますか？ もしそれが出来れば、aboslute_pathに引数を渡すあたりの処理が省略できて、ソースの複雑さが少し下がるように思いました。 :eyes: 

```ruby
# 処理に entry の登場回数が多いので、もうちょっとうまくまとめられそうな気もしますね…
def format_status_with_l_option(entry = self)
  entry.status.format(entry.path, entry.absolute_path)
end
```

heavy-stone replied:
dacf979 にて対応しました！（すみません、コミットメッセージを間違えたままプッシュしてしまったようです🙇‍♂️）

betachelsea commented on 07.ls_object/lib/ls_command.rb:
```diff
@@ -0,0 +1,38 @@
+  def self.exec(options = NO_OPTIONS, paths = [Ls::CURRENT_DIRECTORY])
+    @options = options
+    @has_two_or_more_path = paths.length >= 2
+    new(paths).exec
```
[nits] `exec` がクラスメソッド・インスタンスメソッドの2種類があるので、やや把握が難しいなと感じました :eyes: LsCommandに インスタンスを作る役割 / クラスメソッドを取りまとめる役割、の2つが同居している様子を感じました。見ていてLsCommandにはあまりインスタンスをつく役割としての活躍は求められていないように感じたので、思い切ってクラスメソッドの取りまとめ担当として割り切ってしまう( initialize, exec メソッドどちらも結局中身1行だけなので、 `self.exec()` の中に展開してしまっておく ) のもアリかなと思いました 👀

heavy-stone replied:
こちらはシングルトンでもなさそうでインスタンス化する必要もなく、クラスメソッドだけ使いたいクラスの使い方が分かりませんでしたので、一旦[OOP版のボウリングスコア計算プログラム](https://bootcamp.fjord.jp/products/5130)のコードを参考にさせて頂きました🙏
ただ、インスタンス化する必要がない場合は、クラスメソッドのみの方がすっきりするのでいいのでしょうか？

> ( initialize, exec メソッドどちらも結局中身1行だけなので、 self.exec() の中に展開してしまっておく ) のもアリかなと思いました 👀

841e10e にて対応しましたが、EntryManagerもLsCommandと同じようにインスタンス化せず、クラスメソッドを使う方式に変更した方がよいでしょうか？

betachelsea replied:
なるほど、参考先に沿って作られていたのですね！  すみません、全体の構成をどうするかはかなり個々人の好みが出るところかなと思います。自分はこうなったほうが分かりやすいなと思ったことを書かせていただきましたが 🙇、  heavy-stoneさんがどちらが分かりやすいか、で決めていただいて大丈夫です！

heavy-stone replied:
参照コードを見てインスタンス化が必要ないクラスでなぜnewをしているのかが自分的には分かりませんでしたので、クラスメソッドで実行に変更しました！
174dd03 にて対応しました！

betachelsea commented on 07.ls_object/lib/entry.rb:
```diff
@@ -0,0 +1,107 @@
+
+  attr_reader :path, :parent_path, :is_parent, :path_full_width_count, :path_width, :not_exist, :status
+
+  def initialize(path, parent_path = nil, is_parent: true)
```
[IMO] クラス構成と `is_parent` 周りを拝見しました。自己参照的に親/子関係を持っている様子なのですね。ちょっと構成としては難しめのカタチだな、と感じました :eyes:  こちらですが、親としてのEntryはあくまで1つだけで、子としてさらに `Entry` がいる、という解釈で良いでしょうか？ この時、「親としてのEntry」「子としてのEntry」は、それぞれ役割が明確に違うように感じました。

- is_parent が true のEntry (親) … 子データの取りまとめ
- is_parent が false のEntry (子) … 本質的には entry.status の情報を持ってくること? (事実上同一なようにも思います)
- ( 孫データが存在することはありえない？ )

実行時にpathが2つ以上渡されるケース ( `$ ./ls.rb /no_exist1 /no_exist2 /tmp /usr`)、で、おそらく親データがそのpathの数だけ(親となるフォルダの数だけ?) 発生するのかなと思い、その場合 親 / 子 という考え方は便利かなとは思うのですが、オブジェクトとして 親 = 子 を同一視はしなくても良いのではないか、と自分は思いました。

ただ、こちら個々人のモノの捉え方、見方 といったことが大きく影響するところがあるので、自分の意見だけだと一概には言えないところはあるなとも思います。再検討した結果、いまの状況がベストであるという判断になってもかまわないと考えます。ですが一回、この Entryクラスの役割について再考してみていただいても良いでしょうか？  🙏 「親 / 子 で同一クラスを使用するメリット・デメリット」「親 / 子で明確にクラスを分けたらどうなるか」、一度試してみていただくと面白いかもしれないと思っております！ (自分としては、分けたら、子クラスのほうは EntryStatus に吸収合併されそうな気がしています :eyes: )

heavy-stone replied:
> is_parent が true のEntry (親) … 子データの取りまとめ
is_parent が false のEntry (子) … 本質的には entry.status の情報を持ってくること? (事実上同一なようにも思います)
( 孫データが存在することはありえない？ )

Entry (親)もディレクトリ以外のファイルの場合は自身のファイル情報を出力するため、EntryStatusの情報が必要になるかと思います🙇‍♂️
孫データは存在しない想定です。

> オブジェクトとして 親 = 子 を同一視はしなくても良いのではないか、と自分は思いました。

こちら、当初はEntryとChildEntryにクラスを分けて作成していたのですが、EntryStatusを作成していくうちにEntryとChildEntryの両方に必要となり、差分がなくなってきたので一つにまとめた方がいいのではないかと考えてまとめてしまいました。

heavy-stone replied:
> ですが一回、この Entryクラスの役割について再考してみていただいても良いでしょうか？ 🙏 「親 / 子 で同一クラスを使用するメリット・デメリット」「親 / 子で明確にクラスを分けたらどうなるか」、一度試してみていただくと面白いかもしれないと思っております！

必要最低限の設計で、きれいなコードで早く書くを目標にしているのですが、まだ拡張性を考えたりEntryという意味では親も子も同じという考えをしてしまっていました。
3173181 にてChildEntryを追加して、Entryの自己参照を削除してみましたが、循環しなくなったので、コードの見通しがよくなったかと思います🙏

betachelsea commented on 07.ls_object/lib/entry_manager.rb:
```diff
@@ -0,0 +1,58 @@
+    not_directory_entry_line = format_to_not_directory_line(not_directory_entries)
+    directory_entry_line = format_to_directory_line(directory_entries)
+
+    [error_line, not_directory_entry_line, new_line, directory_entry_line].join
```
[nits] 変数がいっぱい準備されていますね！ 変数名とメソッド名がほぼイコールなものが多いなと感じたので、こういうやり方もありかもと思って共有します 💡 

```ruby
# 変数に取らずに、直接 配列の中身 として扱って join する （ ある意味で ズボラ かもしれませんが… 苦笑 )
[
    format_to_error_line(error_entries),
    format_to_not_directory_line(not_directory_entries),
    has_not_directory_and_directory_entries ? "\n" : '',
    format_to_directory_line(directory_entries),
].join
```

heavy-stone replied:
ありがとうございます！
cff4884 にて対応しました！

betachelsea commented on 07.ls_object/lib/entry_status.rb:
```diff
@@ -0,0 +1,89 @@
+require 'etc'
+
+class EntryStatus
+  FTYPE_BLOCK_SPECIAL = 'blockSpecial'
```
定数が活用されていますね！ 👍

betachelsea commented on 07.ls_object/lib/common_entry_methods.rb:
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+module CommonEntryMethods
+  def update_max_widths(entries)
```
[IMO] Entry, EntryManager ともに同じメソッドが必要になったから書き出された、という経緯かな、と思いました :eyes:   EntryStatus ひとつひとつに max_width が保存してあって、それを上書きするという仕組みなのですね。全体を見て思ったことは、1件のEntryが持つ EntryStatusの `max_〇〇_width` は全て同一の値なのだな、ということです。その場合、Entry自体が1個だけ `max_〇〇_width` を持っていて、それを使うというカタチでもありなのではないか、と感じたのですが、いかがでしょうか？

heavy-stone replied:
こちら、Entryでmax_widthを持つ方法も検討したのですが、その時はStatusに関連する情報なので、Statusが持っていた方がいいと考えて全てのmax_widthをStatusに持たせていたと思います🙇‍♂️
また、EntryManagerとEntryがmax_widthを持つことになり、複雑になりそうだったのでStatusに持たせた方が簡単かと思い実装していました。

005ad17 0b7059e にて対応しました！

betachelsea commented on 07.ls_object/lib/entry_status.rb:
```diff
@@ -0,0 +1,89 @@
+  def format_mode(ftype, mode)
+    rwx_mode = format_to_rwx_mode(mode)
+    case ftype
+    when FTYPE_BLOCK_SPECIAL then "b#{rwx_mode}"
+    when FTYPE_CHARACTER_SPECIAL then "c#{rwx_mode}"
+    when FTYPE_DIRECTORY then "d#{rwx_mode}"
+    when FTYPE_LINK then "l#{rwx_mode}"
+    when FTYPE_FIFO then "p#{rwx_mode}"
+    when FTYPE_SOCKET then "s#{rwx_mode}"
+    else "-#{rwx_mode}" # macのls -lでは'whiteout'があるが、ftypeには含まれず、'unknown'などの場合は'file'としておくため
```
[nits] この辺りの法則性としては、 ftypeの値に対して 先頭の英字が変わる + rwx_mode 、の表記にする、という感じのようですね。すでに `OCTAL_CHAR_TO_RWX` で実施されているのと同じ方法で、もしかするとこの辺りの書き方はグッと減るかもと思いました :eyes: 

```ruby
FTYPES = {
  'blockSpecial' => 'b',
  'characterSpecial' => 'c',
   # ...
}
#=> "#{FTYPES[ftype] || '-'}#{rwx_mode}"
```

heavy-stone replied:
ありがとうございます！
d4cf680 にて対応しました！

betachelsea commented on 07.ls_object/test/l_option_test.rb:
```diff
@@ -0,0 +1,48 @@
+  def test_l_option_with_multiple_paths
+    paths = ['/usr', '/dev/disk1', '/dev/zero']
+    expected = <<~TEXT
+      brw-r-----  1 root  operator  0x1000004 May  7 10:15 /dev/disk1
```
テスト完備なんですね！ すごいですね 👍 ✨ 

[nits] 強いて言えばですが、おそらくこの実行結果は heavy-stone さんの環境にて実施したものをテストされたのかなと思いました。多分自分(レビュアー)の環境で動かしてもテストが通らなさそうです。今回は lsコマンド ということでテストの難易度が高いので止むを得ないとは思うのですが、もしテストを準備するのであれば、テストに用いるサンプルファイルを一緒にコミットしてしまって、他者の環境で clone された場合でも、テストがpassするような状態にしておくのが理想的です。参考まで！ 🙏

heavy-stone replied:
上記についてテスト方法に迷い、[こちらの日報](https://bootcamp.fjord.jp/reports/98419#%E4%BB%8A%E6%97%A5%E3%81%AF%E4%BD%95%E3%82%92%E3%82%84%E3%81%A3%E3%81%A6%E3%81%A9%E3%81%86%E6%84%9F%E3%81%98%E3%81%9F%EF%BC%9F)で記載しましたが、やはりテスト用のサンプルフォルダを用意した方がいいとのことですね🙇‍♂️
実際、コミット時に/tmpのファイル内容が変わってテストを修正することになったりしました( 826c1e6 )ので、今後のプラクティスではサンプルフォルダにファイルを入れるようにしてみます🙏

（再度テストすると/tmpディレクトリ内のファイルが戻っていたので削除しました・・・ 5244349 ）

betachelsea reviewed (commented):
提出お疲れ様です！ コメントしました〜 ✍️

heavy-stone committed:
fix updated test result by local env

heavy-stone committed:
fix updated test result in local env

heavy-stone committed:
remove instance methods

heavy-stone committed:
refactor output method

heavy-stone committed:
update handling max_widths

heavy-stone committed:
update FTYPE constants

heavy-stone committed:
remove binding.irb

heavy-stone committed:
fix updated test result by local env

heavy-stone committed:
remove unused variables

heavy-stone committed:
add ChildEntry class and remove self ref from Entry

betachelsea commented on 07.ls_object/lib/entry_status.rb:
```diff
@@ -0,0 +1,78 @@
+    "#{FTYPES[ftype] || '-'}#{rwx_mode}" # macのls -lでは'whiteout'があるが、ftypeには含まれず、'unknown'などの場合は'file'(-)としておく
+  end
+
+  def format_to_rwx_mode(mode)
```
[nits] `format_to_rwx_mode` の使いどころとしては、`format_mode()` から呼び出されるのみのようですね。かつ、そこまで大規模な処理でもないので、この2メソッドは 1つの手続き として1メソッドにまとまっていても良いなと思いました :eyes:

heavy-stone replied:
コード修正でコード量が小さくなりましたので、まとめた方がいいですね！
325ce34 にて対応しました！

betachelsea commented on 07.ls_object/lib/entry_manager.rb:
```diff
@@ -0,0 +1,60 @@
+    not_directory_entries, directory_entries = valid_entries.partition(&:not_directory?)
+    has_not_directory_and_directory_entries = !not_directory_entries.empty? && !directory_entries.empty?
+
+    not_directory_entries.reverse! if LsCommand.option_r?
```
[IMO] 改めて読んでいて不思議だな、と思ったことを共有しておきます。このEntryManagerは LsCommand から呼び出されているはずなのに、引数などで値を渡すこと無しに、こうして `LsCommand.option_r?` が正常に機能しているというのはちょっと驚きました。むしろ LsCommandのinitialize が走る  → そこで EntryMangerの機能が呼び出される、という順番が確実に守られているからこそ `LsCommand.option_r?` がちゃんと機能していると解釈しました。その順番を無視して、 `LsCommand.option_r?` が先に呼ばれた場合、 nil が返ってくると思いました。子から親を呼び出すということは今回の実装のようにあり得ることではあるのですが、一般的には引数とかで何かしら情報が渡されたものを活用するというのがよくあるパターンかなと思います。

今回のケースは、プログラム全体として、順番が正しく守られているからこそ動くようなカタチになっているなと感じました。もっと巨大なプログラムで、クラスがどんどん使いまわされるような世界観になってくると、もうちょっと「順番がどう呼ばれても問題ないように」「むしろ順番が必要なものはひとまとまりにする」あるいは「インスタンス化して引数として渡してあげる」ということが求められてくるかもしれません 👀  感想でした！ 🙏 現状では動いているし、お気持ちわからなくもないので、 LsCommandの使い方としては現状でOKと思います。

heavy-stone replied:
> 子から親を呼び出すということは今回の実装のようにあり得ることではあるのですが、一般的には引数とかで何かしら情報が渡されたものを活用するというのがよくあるパターンかなと思います。

こちらは、親子ではなくグローバルな状態管理者にアクセスしてオプションが設定されているか確認しているという考え方でした🙇‍♂️
当初はOptionHandlerというモジュールを作成してオプションを管理するようにしていたのですが、パス数も管理が必要になり、LsCommandという名前的にこちらにつけてしまっていてもいいのではということで統合していました。
LsCommandクラスとOptionHandlerといったグローバル状態管理クラス（モジュール？）という別々のクラスに分けた方がいいでしょうか？

> 「順番がどう呼ばれても問題ないように」「むしろ順番が必要なものはひとまとまりにする」あるいは「インスタンス化して引数として渡してあげる」ということが求められてくるかもしれません 👀 感想でした！ 

すみません、こちらはLsCommandが作成された後にEntryManagerが作成される流れになっていると思いましたので、「むしろ順番が必要なものはひとまとまりにする」の方に該当するかと思いましたがいかがでしょうか🙇‍♂️
EntryManagerが呼ばれる前にオプションの状態はどこかに保持しておかなければ参照できないと思いましたので、「順番がどう呼ばれても問題ないように」の例は思い付きませでした😔
もし、何かいい例があれば教えて頂けると助かります🙏

betachelsea replied:
> EntryManagerが呼ばれる前にオプションの状態はどこかに保持しておかなければ参照できないと思いましたので、「順番がどう呼ばれても問題ないように」の例は思い付きませでした😔 もし、何かいい例があれば教えて頂けると助かります🙏

すみません、こちら前回お返事漏れていましたね！ :bow: 

自分が気になっているところは、EntryManagerは次のように使うことはできない、というところです。

```ruby
paths = ['.']
EntryManager.format_entry_group(paths)
```

上記を素直に実行すると、LsCommandは未実行なので `LsCommand.option_l?` とかは nil になってしまうと思います。そのため、今は LsCommand 経由でのみ使われるものとして扱われている、というところに 順番が…という表現をしました。ちょっと表現に語弊がありましたよね…すみません！ :bow: 
こちらの解決ですが、引数で渡してしまうのは一つのアイデアかなと思います。その時にクラスごと渡すのに違和感があれば、new してインスタンス作っても良いかなと思います。


```ruby
ls_command = LsCommand.new # ※現在の想定とやや使い勝手が違うので命名は要再検討と思いますが
EntryManager.format_entry_group(paths, ls_command) # 引数でもらうようにしておけば、絶対渡す前に動かしてることになる
# どちらにせよ LsCommandクラスと EntryMangaerクラスのどちらも必要ということに変わりはないが、
# EntryManagerを単独で呼び出そうとしても呼び出せない状況を作ったほうが、他の実装者が確実にクラスを意図通りに使用することにつながる。
```

現在の LsCommand, EntryManager の関係は、お互い強く結合している状況かなと思います。LsCommandを呼び出す時には良いのですが、EntryManagerをひとりで動かそうとすると、動かせそうにみえて実は問題がある状況のようです。今回の用途としては、密結合していることに特に不利益は無いといえば無いのですが、オブジェクト指向を考慮して進める場合は、疎結合な状態のほうがよりシンプルで使いやすいのでオススメしたいです :bulb: 

また、`LsCommand.two_or_more_paths?` もプログラム内においてかなりイレギュラーな存在のように思います。使い所としてはEntryGroupの `フォルダ名: ` という表示を出すか出さないかという判定のためだけのようなので、EntryGroupに「フォルダ名を出す/出さないの判断フラグ」として引数で指定するのもありかなと思いました。

heavy-stone replied:
EntryManagerのみでも使えるように見えますが使おうとすると、LsCommandのoptionsが初期化されていないので使えないということですね🙇‍♂️
また、EntryManagerを使う際にLsCommandを引数で与えることで、両クラス間の依存を読み手に伝えることができるということですね。

上記より、やはりOptionHandlerクラスがあった方がいいかと思い、 c64b2ff にて対応しました！
インスタンス化する必要はない気はしますが、初期化処理が必要という意図を伝えるならインスタンス化した方がいいのかと思い、今回はnewするようにしました！

ただ、OptionHandlerがバケツリレーのように渡されている点が気になりました。
ですが、[こちら](https://github.com/heavy-stone/ruby-practices/pull/11#discussion_r1696176627)でコメントしたようにグローバルな状態管理クラスを作ってしまうと、元のコードと同じくnew時にOptionHandlerのインスタンスを渡せないので依存関係が分からなくなりますね・・・

betachelsea replied:
はい、そうです！ >  EntryManagerのみでも使えるように見えますが〜
言語化ありがとうございます！説明がつたなくて申し訳ないです 🙇 

> OptionHandlerがバケツリレーのように渡されている点が気になりました。

確かに気になりますね。こういう時は全体を俯瞰してリファクタリングするチャンスかな、と思いました！  :bulb: 
このような観点から見直しを試してほしい、ということを別途まとめてコメントしますね！

betachelsea commented on 07.ls_object/lib/entry.rb:
```diff
@@ -0,0 +1,100 @@
+
+  def initialize(path)
+    @path = path
+    @path_full_width_count = @path.scan(FULL_WIDTH_REGEX).count
+    @path_width = @path.length + @path_full_width_count
```
[MUST] もしかして:  `@path_full_width_count` や `@path_width` の呼び出しをみたところ、 ChildEntry からしか呼ばれていないようなので、 Entryは持ってなくても大丈夫だったりしませんか？ 再確認してみてほしいです 🙏

heavy-stone replied:
確かにそうですね、自己参照時のものが残ったままだったようですので削除します🙇‍♂️
d7fa064 にて対応しました！

betachelsea commented on 07.ls_object/lib/entry_manager.rb:
```diff
@@ -0,0 +1,60 @@
+    if LsCommand.option_l?
+      not_directory_max_widths = calc_status_max_widths(not_directory_entries)
+      not_directory_entries.map do |entry|
+        entry.format_status_with_l_option(not_directory_max_widths)
```
なぜ Entry と ChildEntry がどちらも `format_status_with_l_option` を持っているんだろう、似たような機能なのにどちらも持っていてメンテが大変だな、とおもったら、こちらで使っていたのですね :eyes: 次のような呼び出しの時に必要ということですね。

```console
$ ls -l aaa.txt bbb.txt ccc.txt
```

改めて全体をみて、Entry と ChildEntry の責務がやや重なっているところがあるなと感じています。そこをもっと綺麗に分けられれば、もっと関係性をスッキリできるのではないかと思います。
「処理階層」で考えるのではなく、「表示の役割」的なところで区分できないでしょうか？ 

この `format_status_with_l_option` をEntryに持たせるのをやめて、ChildEntryが全部持たせることができるのではと思います。not_directory のファイルたちを、仮想的に Entry のまとまりとして扱ってやってみるのはどうでしょうか。totalの表示がない、フォルダ表示がない、というところは 通常の「フォルダ」扱いとは分けてあげる必要があるかもしれませんが、パラメータで区別できるような気がします。

画像でイメージしてみましたが…どうでしょう？

### 現在
![sample_1](https://github.com/user-attachments/assets/bb2a38bc-73eb-46ea-a38b-ef5e117250ea)

### 提案

こう考えると、機能を完全に分けて、役割分担できるのではと思いました。(もしかすると、クラスの命名を考え直したくなるような提案かもしれません 🙇  )

![sample_2](https://github.com/user-attachments/assets/43d6d057-3e46-4a94-88a0-cc3ea12cd378)
![sample_3](https://github.com/user-attachments/assets/6ab38b3c-f911-4a40-a925-c915b740aa2a)
![sample_4](https://github.com/user-attachments/assets/bb8659a4-7757-4da5-9523-a036b1777dd0)

heavy-stone replied:
画像を交えてのご説明ありがとうございます！
当初、表示出力を見てクラス設計をしていったのですが、エラーエントリやディレクトリでないエントリ群をクラス管理しやすくグループ化するというところまで頭がまわりませんでした🙇‍♂️
上記提案を踏まえて、下記にクラス名を変更して 7d69553 にて対応しました！
Entry → EntryGroup
ChildEntry → Entry

betachelsea reviewed (commented):
画像で出してみました！ ピンとこなかったら、もうちょっとコメントやり取りで煮詰めていければと思います 🙇 ちょっとチェックしてみていただきたいです！

heavy-stone committed:
remove unused variables

heavy-stone committed:
unify methods

heavy-stone committed:
use EntryManager class method instead of instance method

heavy-stone committed:
update class structure

betachelsea commented on 07.ls_object/lib/entry_manager.rb:
```diff
@@ -0,0 +1,63 @@
+      error_entry_group.format_error_entries,
+      not_directory_entry_group.format_not_directory_entries,
+      has_not_directory_and_directory ? "\n" : '',
+      directory_entry_groups.map(&:format_directory_entries).join("\n")
```
このあたり、4パターンが準備されていて、それぞれ呼び出しているのですね。
仕組みが多くてちょっと難しさを感じています 👀  手続きの処理にそれぞれ名前がついて呼び出していますが、概念を取りまとめて、同じような形で呼び出せたらスッキリするかもと思いました。

[MUST] 次の対応を検討してみてもらえますか？ 🙏 
- `format_error_entries` / `format_not_directory_entires` / `format_directory_entires` の呼び出し分岐の判別は `EntryGroup` にまかせる。`EntryGroup` は自身がどんなTypeなのかを知っているので可能かと思います。
-  has_not_directory_and_directory で `"\n"`を出すか出さざるか、というところはここでわざわざ判別なくてもうまくできる方法を考えたいな、と感じます。例えば not_directoryのほうの出力結果を nil にしてあげるとか…? 

```ruby
# 雰囲気だけですがこんな感じになったらいいなと思います
entry_groups = [
  EntryGroup.new(error_paths, ENTRY_GROUP_TYPE[:error]),
  EntryGroup.new(not_directory_paths, ENTRY_GROUP_TYPE[:not_directory]),
  **create_directory_entry_groups(directory_paths), # ちょっとこう書けるかは不明ですが…
]

entry_groups.map(&:format_*****).join("\n")
```

また、こうすると条件判断が EntryGroupの中にまとまってきそうです。
EntryGroupの中では、いま、上記の directory の様子 と掛け合わせで l_option の有無で大きな分岐毎のメソッドが作られていますが、うまく取りまとめできないでしょうか？ できればかなり省エネなコード量になると思っています。

heavy-stone replied:
> format_error_entries / format_not_directory_entires / format_directory_entires の呼び出し分岐の判別は EntryGroup にまかせる。EntryGroup は自身がどんなTypeなのかを知っているので可能かと思います。

提出時はEntryGroupに一つのメソッドで判断させていましたが、分岐がEntryManagerと同じで冗長と考え、それぞれのメソッドに分岐させました。
ですが今回、group_typeを追加したことで以前よりシンプルに分岐できるようになったので、 8bb4d9c で対応しました！

> has_not_directory_and_directory で "\n"を出すか出さざるか、というところはここでわざわざ判別なくてもうまくできる方法を考えたいな、と感じます。例えば not_directoryのほうの出力結果を nil にしてあげるとか…?

すみません、こちらは下記の画像のようにnot_directoryとdirectoryのあるなしによって改行を判断する必要があり、それができるのはEntryGroupの一階層上のEntryManagerだと思いますので、has_not_directory_and_directoryを使わずに実現する方法は分かりませんでした🙇‍♂️
また、出力結果をnilにするというのは、現状not_directoryがない場合は空文字列を返すようにしていますので差はないかと思いましたが、思い違いがありましたらご指摘お願いします🙏
![has_not_directory_and_directory](https://github.com/user-attachments/assets/fd028dae-6353-433d-9d2b-70c8444ccc76)

コメント頂いた通り、スプラット演算子を使うとEntryGroup配列に対してmapでformat_****メソッドをループするシンプルな書き方ができるかと思いますが、改行判断の必要により一旦元のコードのままとしています🙇‍♂️

betachelsea replied:
改行の仕様のためにコードが読みづらいのは、なんだか勿体無いな〜と思います :eyes:  特に、今回のケースは法則性を見つければ解決可能なように思いました。また、読み手としては、改行の判断の処理がかなり複雑に感じており、できれば `has_not_directory_and_directory` は無くしたいと思いました。
次のようにすれば改行問題は解決すると思いますが、どうでしょう？

- (1). EntryGroup の `format_〇〇` で `return '' if @entries.empty?` を実行しているが、これを `return nil if @entries.empty?` とする。
- (2). format_〇〇 全呼び出しの時↓に、 join の前に compact を実行して nil排除し、joinは ("\n")を使って結合させる。
- (3). EntryGroup の `format_error_group` メソッド内で実行されている  `concat("\n")` が、(2)の join("\n") と同様の役割を持っており重複しているので、削除する。


```ruby
# (2) はこんな感じのイメージです
[
    error_entry_group.format_entry_group,
    not_directory_entry_group.format_entry_group,
    directory_entry_groups.map(&:format_entry_group), # ここも 改行は最後に任せる
].compact.join("\n")

# なぜ nil を返して compact するのか?  →空白 "" や nil も1要素として join("\n")するときに対象になるので、排除してからjoin("\n")したかった。そして、配列の中から空文字列( `""` )を探し出して除外するよりも、nilを除外するほうが メソッドが用意されており(compact)、簡単だから。
```

heavy-stone replied:
ありがとうございます！
上記参考に修正してみてあと一歩のところまで行ったのですが、エラーエントリのみの場合に最後の改行がない状態となってしまいました🙇‍♂️

テストとエラーエントリのみ時の出力結果
![test_error_entry](https://github.com/user-attachments/assets/44851ab3-3202-40ee-8018-dd7ca40223b6)
エラーエントリ時は末尾に改行をつけておき、最終的に改行が2つある場合は1つにする方法も考えましたが、逆に分かりにくいかと思い、一旦元のコードのままとしています🙇‍♂️

変更差分
![entry_group](https://github.com/user-attachments/assets/52ae811a-2a66-441a-99eb-29a0a4418519)
![entry_manager](https://github.com/user-attachments/assets/4eb03f65-90bb-4312-8107-4ac56da5037d)

betachelsea replied:
おっと、すみません、自分の環境だと通常lsも最後改行無かったので、てっきり行けるかと… 😭  

![スクリーンショット 2024-08-05 23 31 35](https://github.com/user-attachments/assets/149ebb49-d4bc-4699-9d80-f8a7fb8834bb)


その部分が改行必要であれば、パターン化は結構厳しそうですね。そう考えると、その改行の有無は確かに EntryGroup群を統括する EntryManagerの役割、と断じても良さそうです 🙆

heavy-stone replied:
Preztoのpureというテーマを使っているのですが、このテーマは最終行に改行を加えるようです。
頂いた画像と同じテーマと思われるsorinに変更すると通常のlsと作成したlsコマンドの出力が一致したので、pureテーマが悪さをしている可能性があると考えて調査すると下記の記事を見つけました！
https://michimani.net/post/develop-zsh-prompt-remove-last-line/

こちらの対応をすると、pureテーマでエラーエントリのみの場合でも最後の改行が表示されるようになりましたので、ご提案頂いたコードで実装できるようになりました！
fb20a49 にて対応しました！
![prezto_pure](https://github.com/user-attachments/assets/bb523744-cb26-47ce-aeb1-a1365f215b63)

betachelsea commented on 07.ls_object/lib/entry_group.rb:
```diff
@@ -0,0 +1,84 @@
+  def format_with_l_option
+    total_block = @entries.sum { |entry| entry.status.block }
+    total = "total #{total_block}\n"
+    formatted_entry_statuses = @entries.empty? ? '' : format_entry_statuses_with_l_option
```
[MUST] `format_entry_statuses_with_l_option` の呼ばれ方ですが、次のようになっているようです。

- format_not_directory_entries
  - lオプションあれば → format_entry_statuses_with_l_option
- format_directory_entries
  - lオプションあれば → format_with_l_option
    - entriesがあれば → format_entry_statuses_with_l_option

これらの処理はたくさん分岐がありますが、もっとコンパクトにまとめられるのではないかと思います。
例えば、`format_with_l_option` で表示される total_blockの有無は、その前段階の not_directory OR directory というところで決まります。
なので、1メソッドの中で、
- directoryがあるなら、directoryの情報を表示する
- lオプションがあれば、entires情報を表示する

といった表現ができるように思います。整理していくともうちょっと違った形になるかもしれませんが…。

現在、特定条件のためのメソッド、という形になっている箇所が多く、分岐が多い分処理を追っていく難易度が少し高い状況にあると感じます。うまくそれらをオブジェクト指向的な要素を活用して、解消にトライしていってほしいです！オブジェクト指向では、色んなステータスをオブジェクト自身が知っています。ステータスに応じて外側から呼ばれるものを変えるのではなく、自分自身がどういう状態なのかをうまくチェックしながら処理してみてもらいたいです。

heavy-stone replied:
> 現在、特定条件のためのメソッド、という形になっている箇所が多く、分岐が多い分処理を追っていく難易度が少し高い状況にあると感じます。うまくそれらをオブジェクト指向的な要素を活用して、解消にトライしていってほしいです！

コードが読みにくいことは感じていたのですが、うまく解消する方法が分かりませんでした。
[こちらで頂いたコメント](https://github.com/heavy-stone/ruby-practices/pull/11#discussion_r1695417304)を参考に、出力される項目を表示される順番通りにコードの分岐として書いていくことで見通しは良くなったのかと思いました🙏
今回、「オブジェクト指向的な要素を活用」はTYPES定数でちょっとは実践できたかと思いますが、十分かは分かりません🙇‍♂️
上記対応と合わせて 8bb4d9c にて対応しました！

betachelsea reviewed (commented):
修正ありがとうございます！大きな観点のところから、コメント追加させていただきました 🙏

heavy-stone committed:
add format_entry_group and improve readability

heavy-stone committed:
fix rubocop warnings

betachelsea reviewed (commented):
拝見しました！ :eyes:

heavy-stone committed:
add OptionHandler

betachelsea reviewed (commented):
バケツリレー化になっている、という問題より、改めて全体を俯瞰すると、さらにオブジェクト指向化を進めていけそうなポイントが見えてきたように思います。次のような観点から、見直しを検討してみていただきたいです！ オブジェクトはそれぞれ何を知っていればいいのか？ということをどんどん分解していく・名前をつけていく・シンプルにする工夫をしていくのが、オブジェクト指向の醍醐味だと思います！

## 検討していただきたいこと

- `show_directory_name?` について
  - こちら、`two_or_more_paths?` がリネームしたのですね！とても素晴らしい名前になったと思います :raised_hands: 
  - これはOptionHandlerの役目なのか？ EntryManagerが単独で判断できるのではないか？ と思いました。OptionHandlerがちゃんと名前の通りOptionのことしか考えなくて済むようにしてあげたいです。
- EntryManager内部でのバケツリレーについて
  - このバケツリレーの様子を見ると、改めてEntryManagerをインスタンスとして扱う価値が出てきたかも、と思いました。インスタンス変数経由で option_handler を扱えると、だいぶ楽になる気がしました。
  - 改めてEntryManagerをインスタンスとして考えると、今の `self.format_entry_groups()` が initalizer と 出力の2つの機能が同居しているように思えてきました。うまく分離できると読みやすくなりそうと思いました。
    - インスタンス化するならば、has_not_directory_and_directory は メソッド化したいと思いました。また、これも現時点で判定そのまんまの変数名なので、何を判断したいのかに着目した `show_directory_name?` のようなリネームが望ましいとも思いました。
  - インスタンス化すると `EntryManager.new(...).〇〇` みたいな感じで newしたあとにすぐ呼んで終わり、みたいな感じになってやや悲しいですが、致し方ないかなとも思います。それを違和感なく読ませるために 当初参考にしていただいた OOP版のボウリングスコア計算プログラム のような実装がありそうな気がしてきました。すみません、なんだか一周回ってきましたね :bow:  (※→構成案見出し参照)
- EntryGroupに本当に option_handlerを渡す必要があるのか？
  - ここもバケツリレーはあまり望ましくなさそうですね。
  - option_handler の使いどころをみると、 `option_l?` の判断 OR `show_directory_name?` の判断をしているようです。全てに用事がないのであれば、EntryGroupに「自分は lオプションの表示をする/しない」「自分は directory_name を表示する/しない」というフラグを教えてあげて、表示させる、というやり方もありかなと思いました。引数は増えますが、バケツリレーするよりもシンプルかなと思いますがどうでしょう？
    
  ### ※構成案
  
  ```ruby
  # 案(注: 完全にOOP版ボウリングスコア計算プログラムの構成をマネしました)
  # https://bootcamp.fjord.jp/products/5130
  
  class EntryManager
    def self.format_entry_groups(paths, option_handler)
      new(paths, option_handler).format_entry_groups
    end
    
    def initialize(paths, option_handler)
      # ...
    end
    
    def format_entry_groups
      # ...
    end
    
    private
    
    # ... 
  end
  
  # ------
  EntryManger.format_entry_groups(paths, option_hander) #=> 出力結果が得られる
  ```

heavy-stone committed:
remove has_not_directory_and_directory variable

modify prezto output by setting .zshrc
https://michimani.net/post/develop-zsh-prompt-remove-last-line/

heavy-stone committed:
add is_directory_name_shown argument instead of show_directory_name? method

heavy-stone committed:
add is_status_shown argument

heavy-stone committed:
refactor EntryManager

heavy-stone commented:
> show_directory_name? について
>これはOptionHandlerの役目なのか？ EntryManagerが単独で判断できるのではないか？ と思いました。

単独判断できそうですので、 2f7d089 にて対応しました！

> EntryGroupに本当に option_handlerを渡す必要があるのか？
>全てに用事がないのであれば、EntryGroupに「自分は lオプションの表示をする/しない」「自分は directory_name を表示する/しない」というフラグを教えてあげて、表示させる、というやり方もありかなと思いました。

そうですね、要不要を判断してそのクラスに必要な情報のみ与えるのがいいかと思いましたので、 8b6ef56 にて対応しました！

>EntryManager内部でのバケツリレーについて
>改めてEntryManagerをインスタンスとして考えると、今の self.format_entry_groups() が initalizer と 出力の2つの機能が同居しているように思えてきました。うまく分離できると読みやすくなりそうと思いました。

すみません、分離すると読みやすくなるというのはわかりますが、インスタンス化する理由についてよろしければ教えて頂きたいです🙇‍♂️
初期化処理とそれ以外の処理が必要な場合はインスタンス化した方がいいということでしょうか？
例えば、self.format_entry_groups()内でプライベートクラスメソッドでEntryGroup作成メソッドと出力メソッドという風にメソッドを分けてしまえば読みやすさは高まるかと思いました。
また、複数インスタンス化する必要がないクラスはクラスメソッドのみで定義した方がいいのでは考えてしまっていますが、こちらをインスタンス化するとLsクラスもインスタンス化する必要があるのではないかと考えています。
一旦、 018aa2b にて対応しましたが複数インスタンス化できないようにした方がいいのではないかと思い、シングルトンパターンとしています。

- OptionHandlerクラスの削除
「EntryGroupに本当に option_handlerを渡す必要があるのか？」の対応で、EntryGroupにOptionHandlerを渡す必要がなくなり、EntryManagerのみに渡す構成になりました。
元々OptionHandlerを導入した意図としては複数クラスで参照される場合にオプションを管理するクラスがあると便利ということでしたが、現状OptionHandlerの処理が小さくなりEntryManager自体が渡されたオプションを見て対応するので十分かと思い、OptionHandlerは削除しました🙇‍♂️
- LsComandクラスの縮小化
リファクタリングした結果、LsComandの処理がほぼなくなったようにも見えますので、EntryManagerの中身をLsComandに移してEntryManagerクラスは削除してもいいのではないかと考えています。

betachelsea commented on 07.ls_object/lib/entry_manager.rb:
```diff
@@ -0,0 +1,78 @@
+
+    reverse_paths_if_needed!(not_directory_paths, directory_paths)
+
+    @error_entry_group = EntryGroup.new(error_paths, EntryGroup::TYPES[:error])
+    @not_directory_entry_group = EntryGroup.new(not_directory_paths, EntryGroup::TYPES[:not_directory], is_status_shown: option_l?)
+    @directory_entry_groups = create_directory_entry_groups(directory_paths)
```
[nits] これらのインスタンス変数はそれぞれ `format_entry_groups` からしか呼ばれていないので、そちらのメソッドにお引越ししてローカル変数にしたら、大きなインスタンス変数が減って管理が楽になるかもと思いました 💡  代わりに `error_paths` などをインスタンス変数として扱ったら 汎用的かなと思います。

heavy-stone replied:
すみません、気になったのですが初期化という意味はattributes (属性)を設定することで、EntryGroupとの関連については入りませんでしょうか🙇‍♂️
というのは、EntryManager（LsComand）の初期化処理として、関連があるEntryGroupをnewして関係として持つという処理も入るのではないかと考えたからです。

[こちらで](https://github.com/heavy-stone/ruby-practices/pull/11#pullrequestreview-2219326724)下記コメント頂いていました。
> 改めてEntryManagerをインスタンスとして考えると、今の self.format_entry_groups() が initalizer と 出力の2つの機能が同居しているように思えてきました。

initializerとして、initializeメソッドでEntryManager（LsComand）のattributes設定とEntryGroupとの関連を初期化しておき、format_entry_groupsはその初期化した自身の状態を参照して出力処理をするというイメージかと思いましたが、いかがでしょうか？

一旦、 6849e2f にて対応しました！

betachelsea replied:
対応ありがとうございます！

>  initializeメソッドでEntryManager（LsComand）のattributes設定とEntryGroupとの関連を初期化しておき、format_entry_groupsはその初期化した自身の状態を参照して出力処理をするというイメージ

そうですね、`2つの機能が同居` というのは 変数の準備 と 出力が一緒だなと思ったところが大きかったので、`EntryGroup.new` は initialize でも format_entry_groups でもどちらでもいいと思います。完全に format_entry_groups に出力機能だけを任せようとおもったら、ご提案いただいたように `EntryGroup.new` の結果を attributes として扱うのも良さそうですね 🙆‍♂️

betachelsea commented on 07.ls_object/lib/entry_manager.rb:
```diff
@@ -0,0 +1,78 @@
+
+  def initialize(options, paths)
+    @options = options
+    @is_directory_name_shown = paths.length >= 2
```
[nits] インスタンス化することにしたら、`@is_directory_name_shown` はインスタンス変数としての管理ではなく、メソッドとして扱えそうかなと思いました。 インスタンス変数、便利な反面、うっかりどこでも編集しやすいということもあり、結構レビュー側としては注意を払わなければならない存在、というイメージです。できるだけ少ないと、注意点が減って助かる、と感じます。

heavy-stone replied:
インスタンス変数にするかメソッドにするか迷う時はあるのですが、変更されて困るものはインスタンス変数ではなくメソッドとするという着眼点はあまりなかったので助かります🙏
146ad61 にて対応しました！

betachelsea commented on 07.ls_object/lib/entry_manager.rb:
```diff
@@ -0,0 +1,78 @@
+    valid_paths, error_paths = sorted_paths.partition { |path| File.exist?(path) || File.symlink?(path) }
+    directory_paths, not_directory_paths = valid_paths.partition { |path| directory?(path) }
+
+    reverse_paths_if_needed!(not_directory_paths, directory_paths)
```
[FYI] 末尾に `!` ついていますが、この `!` の定義としては、実は次のようになっています。
> def xxx!  「!」はメソッド名の一部です。慣用的に、同名の(! の無い)メソッドに比べてより破壊的な作用をもつメソッド(例: tr と tr!)で使われます。

[Rubyで使われる記号の意味（正規表現の複雑な記号は除く） (Ruby 3.3 リファレンスマニュアル)](https://docs.ruby-lang.org/ja/latest/doc/symref.html)
そう考えると、 `reverse_paths_if_needed` という同一のメソッド名がなく、単独で破壊的なので、このメソッド名には実は `!` をつける必要がない、ということになりそうです。  (自分も最近まで把握できておりませんでした🙇  )

[nits] また、このメソッドの中身としては `option_r?` に応じてひっくり返すかどうか、というところですね。こちらもしかして `sorted_paths` の定義あたりで同時に解決できたりしますでしょうか。呼ばれるのは一度きりというところと、破壊的変更があり、それを名前で表現されないとちょっとややこしい、という点において、メソッドに隠蔽するメリットはあまりないように感じました。

heavy-stone replied:
`!`についての情報ありがとうございます！
`!`(bang)について調査したところ、下記の記述を見つけました。
[Method Names](https://docs.ruby-lang.org/en/3.1/syntax/methods_rdoc.html#:~:text=The%20bang%20methods,other%20Ruby%20libraries.)
> ほとんどの場合、Ruby コア ライブラリには、レシーバを変更しないすべてのバング メソッド (メソッド名が ! で終わる) の非バング対応 (メソッド名が ! で終わらない) があります。

ただ、Railsのソースコードを調査すると下記のように`!`付きメソッドのみのメソッドもいくつかありました😵‍💫
これらは"ほとんど"ではない場合でしょうか・・・
- `actionmailbox/app/models/action_mailbox/inbound_email/message_id.rb`の
`create_and_extract_message_id!`や`create_and_upload_raw_email!`
- `actionpack/lib/action_dispatch/middleware/ssl.rb`の`flag_cookies_as_secure!`

また、プロを目指す人のためのRuby入門を読み返すとP.71のコラムに「破壊的メソッドであっても非破壊的メソッドがない（つまり1種類しかない）場合は!が付かない。」とありましたので、`!`はつけない方がよさそうですね。

そうなると、そもそもメソッド作成時に破壊的に作らなくてよいメソッドはそのように作る方がいいようですね🙇‍♂️
今回で言うと、`reverse!`の代わりに`reverse`メソッドで実装した方がよかったかもしれません。

元々rubocopに行数が多いと警告を受けてメソッド分けた経緯があり、今回`valid_paths.reverse!`に処理をまとめることができたので、 `reverse_paths_if_needed!`自体を 1087580 にて削除しました！

betachelsea replied:
> これらは"ほとんど"ではない場合でしょうか・・・

"ほとんど" ということは、 "すべて" ということでないので、そういうこともあるのですね 😳 
https://github.com/rails/rails/blob/e13b251ae0789e0a4a370ca1c84b4e10d64b643b/actionmailbox/app/models/action_mailbox/inbound_email/message_id.rb#L16

ソースを見てみた印象としては、内部で `create!` を呼んでいるので、その示唆として  `!` をメソッド名に冠しているのかな、と思いました。結構初期からあるメソッドなのか、この命名に関する議論などについてサッとは追いかけることはできませんでした 🙏 
この方法 ( 中で `!` 破壊的メソッドを呼んでいる場合はその外側のメソッド名にも `!` をつけてよい )がTrueならば、先日指摘させていただいた `reverse_paths_if_needed!` もOKということになりそうですね。すみません 🙇 

ですが、調べていただいた通り、ドキュメントでよく書かれているのは `「破壊的メソッドであっても非破壊的メソッドがない（つまり1種類しかない）場合は!が付かない。」` なので、TIPSとして把握しておいていただければと思いました！

今回については `reverse_paths_if_needed!` 自体が非採用になったとのことで、自分としては、変数が `option_r?` を求められているときに変質する様子の見通しが良くなったので、助かりました 🙏

betachelsea reviewed (commented):
修正ありがとうございます！ コメント&内容拝見しました 👀 

> インスタンス化する理由

自分は EntryManager を改めてみた時に、次のように思ったのがインスタンス化ありかな、と思ったキッカケです。
- 持っている `*_paths` 類の変数を attributes (属性) として捉えることができそうだ
- 初期化手続きと出力手続きとでハッキリ分けられそう、initializerがあればそこに初期化を入れるのは読みやすくなる
- 利用にあたって必ず初期化手続きが必要なのであれば、インスタンス化したほうがわかりやすそう

※少しEntryManagerについてこう整理すると良さそう、というのを別途コメントしておきました :pray: 

> こちらをインスタンス化するとLsクラスもインスタンス化する必要があるのではないか

上記と比較して、Lsクラスは単純に「Lsプログラム全体を実行するきっかけ」としてのクラスであり、属性もない・初期化手続きもないので、反対にインスタンス化するメリットは薄そうに思いました。

---

一応、自分が思うインスタンス化する/しない理由を挙げてみましたが、この辺は実装者の好みとか、認知とか、納得感に左右されやすいところだと思うので、heavy-stoneさんが納得される形にしてもらって大丈夫です〜 :pray: 

---

> OptionHandlerクラスの削除

あれ、消しちゃったんですね！ 自分としては役割がハッキリしてるクラスで、結構良いなと思っていました。代替として EntryManagerに `option_a?` などのメソッドが生えましたね。以前はなかったものが追加されたことで、ちょっとEntryManagerの役割がブレてしまうのではとやや懸念しましたが、もし最初からそうだったと言われれば、 これもこれでアリかな…という気持ちで拝見しました。承知しました :ok_man: 


> EntryManagerの中身をLsComandに移してEntryManagerクラスは削除してもいいのではないかと考えています。

なるほど、それも一手かな、と思いました！ :ok_man: EntryManagerが `option_a?` を持っているよりも、 LsCommandという名前のクラスが `option_a?` の判断を持っていた方が、なんだかしっくりくる気がします。

heavy-stone committed:
remove EntryManager to unify with LsCommand

heavy-stone committed:
refactor variables

heavy-stone committed:
add directory_name_shown? method

heavy-stone committed:
remove reverse_paths_if_needed! method

heavy-stone commented:
de08640 にてEntryManagerをLsCommandに統合しました！

インスタンス化する理由についてご記載ありがとうございます🙇‍♂️

> インスタンス化する理由
自分は EntryManager を改めてみた時に、次のように思ったのがインスタンス化ありかな、と思ったキッカケです。
>- 持っている *_paths 類の変数を attributes (属性) として捉えることができそうだ
>- 初期化手続きと出力手続きとでハッキリ分けられそう、initializerがあればそこに初期化を入れるのは読みやすくなる
>- 利用にあたって必ず初期化手続きが必要なのであれば、インスタンス化したほうがわかりやすそう

>上記と比較して、Lsクラスは単純に「Lsプログラム全体を実行するきっかけ」としてのクラスであり、属性もない・初期化手続きもないので、反対にインスタンス化するメリットは薄そうに思いました。

LsクラスとLsCommandクラスはコード量は違いますが、複数インスタンス化が不要で1つの処理をするという点は同じなので、何かしらのルールに従うならクラスメソッドのみにするか、シングルトンインスタンス化するかのどちらかに統一されるのではと考えてしまいます。
Lsクラスは属性としてoptionsとpathsがあるように見え、それをセットすることが初期化で、LsCommand.execを出力するメソッドを別出しできるとも考えてしまいます。
ただ、コード量は違いますので、もしLsクラスの処理が多くなったらインスタンス化を検討していくということかなとも思いましたが・・・コード量によってクラスメソッドかシングルトンインスタンス化を決定する？とも思います😵‍💫

そもそもLsクラスは、テストをしやすいようにLsCommand.execにoptionとpathsを渡せるようにと作成したクラスでしたので、それを踏まえてインスタンス化する必要はないと言えなくもないですが、線引があいまいになりそうです😵‍💫

>一応、自分が思うインスタンス化する/しない理由を挙げてみましたが、この辺は実装者の好みとか、認知とか、納得感に左右されやすいところだと思うので、heavy-stoneさんが納得される形にしてもらって大丈夫です〜 🙏

オブジェクト指向設計は捉え方によってあいまいなところがあるので、その場合は臨機応変にということですね🙏

betachelsea reviewed (approved):
ご意見まとめていただき、ありがとうございます！ LGTMとします！ 🎉 

> オブジェクト指向設計は捉え方によってあいまいなところがあるので、その場合は臨機応変にということですね🙏

そうですね、これからぜひ他の提出物を見ていただきたいと思うのですが、同じことをやろうとしているものでも人によって書き方が違うので、観察して色んな考え方を把握してみていただきたいです。

> クラスメソッドかシングルトンインスタンス化

この辺りも人によって捉え方が違うので、今回レビュアーとして自分の意見を共有させていただきましたが、多分 heavy-stoneさんのポリシーとは違うところもあって、混乱させてしまったかもと思いました。すみません！ 🙇  今回特にイチからのソースコードの構築という点も難しさにあったかなと思います。他のプロジェクトに途中ジョインとかですと、おそらくすでに既存のソースコードがあるので、そこに右ならえする形で構築するのがリポジトリ全体の平和につながっていくと思います。今回のレビュー、大変だったかと思いますが 🙏  以降の糧にしていただけると嬉しいです。

お疲れ様でした〜！ 🎉
